planner/core/plan_cache_lru.go

```diff
@@ -0,0 +1,210 @@
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			l.cache.MoveToFront(element)
+			return element.Value.(*CacheEntry).PlanValue, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *LRUPlanCache) Put(key kvcache.Key, value kvcache.Value, paramTypes []*types.FieldType) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	hash := hack.String(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	// bucket exist
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			element.Value.(*CacheEntry).PlanValue = value
+			l.cache.MoveToFront(element)
+			return
+		}
+	} else {
+		l.buckets[hash] = make(map[*list.Element]struct{})
+	}
+
+	newCacheEntry := &CacheEntry{
+		PlanKey:   key,
+		PlanValue: value,
+	}
+	element := l.cache.PushFront(newCacheEntry)
+	l.buckets[hash][element] = struct{}{}
+	l.size++
+	if l.size > l.capacity {
+		l.RemoveOldest()
+	}
+}
+
+// Delete deletes the multi-values from the LRU Cache.
+func (l *LRUPlanCache) Delete(key kvcache.Key) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	hash := hack.String(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	if bucketExist {
+		for element := range bucket {
+			l.cache.Remove(element)
+			l.size--
+		}
+		l.buckets[hash] = make(map[*list.Element]struct{})
+	}
+}
+
+// DeleteAll deletes all elements from the LRU Cache.
+func (l *LRUPlanCache) DeleteAll() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	for lru := l.cache.Back(); lru != nil; lru = l.cache.Back() {
+		l.cache.Remove(lru)
+		l.size--
+	}
+	l.buckets = make(map[hack.MutableString]map[*list.Element]struct{})
+}
+
+// Size gets the current cache size.
+func (l *LRUPlanCache) Size() int {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	return int(l.size)
+}
+
+// Values return all values in cache.
+func (l *LRUPlanCache) Values() []kvcache.Value {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	values := make([]kvcache.Value, 0, l.cache.Len())
+	for element := l.cache.Front(); element != nil; element = element.Next() {
+		value := element.Value.(*CacheEntry).PlanValue
+		values = append(values, value)
+	}
+	return values
+}
+
+// Keys return all keys in cache.
+func (l *LRUPlanCache) Keys() []kvcache.Key {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	keys := make([]kvcache.Key, 0, len(l.buckets))
+	for _, bucket := range l.buckets {
+		for element := range bucket {
+			keys = append(keys, element.Value.(*CacheEntry).PlanKey)
+			break
+		}
+	}
+	return keys
+}
+
+// SetCapacity sets capacity of the cache.
+func (l *LRUPlanCache) SetCapacity(capacity uint) error {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	if capacity < 1 {
+		return errors.New("capacity of lru cache should be at least 1")
+	}
+	l.capacity = capacity
+	for l.size > l.capacity {
+		l.RemoveOldest()
+	}
+	return nil
+}
+
+// RemoveOldest removes the oldest element from the cache.
+func (l *LRUPlanCache) RemoveOldest() {
+	l.lock.Lock()
```

Comment:
It seems we cannot lock it here, otherwise, there is a deadlock, since `Put` has already locked it and this lock is un-reentrant.